mof_array/pmf/process_mass_data.py

```diff
@@ -276,24 +276,45 @@ def calculate_all_arrays(mof_list, num_mofs, element_pmf_results, gases):
 
     return(mof_array_list, all_array_pmf_results)
 
-def create_bins(mofs_list, calculate_pmf_results, gases, num_bins):
-    """Creates bins for all gases, ranging from the lowest to highest mole fractions for each.
-
-    Keyword arguments:
-    mofs_list -- list of mofs used in analysis
-    calculate_pmf_results -- list of dictionaries including mof, mixture, probability
-    gases -- list of present gases
-    num_bins -- number of bins specified by user in config file
-    """
-    # Creates numpy array of all compositions, needed to calculate min/max of each gas's mole frac.
-    mof = mofs_list[0]
-    temp_one_mof_results = [row for row in calculate_pmf_results if row['MOF'] == mof]
-    comps_array = np.array([[float(row[gas]) for gas in gases] for row in temp_one_mof_results])
-
-    bin_range = np.column_stack([np.linspace(min(comps_array[:,index]), max(comps_array[:,index]) +
-        (max(comps_array[:,index])-min(comps_array[:,index]))/num_bins, num_bins + 1) for index in range(len(gases))])
+# ----- Creates bins for all gases -----
+# Keyword arguments:
+#     gases -- list of present gases
+#     num_bins -- number of bins specified by user in config file
+#     mof_list -- list of mofs used in analysis
+#     element_pmf_results -- list of dictionaries including mof, mixture, probability
+def create_bins(gases, num_bins, mof_list, element_pmf_results):
 
-    bins = [ { gases[index] : row[index] for index in range(len(gases)) } for row in bin_range]
+    # Save list of dictionaries for first MOF to use as a list of all gas mole fractions
+    # Might be good to make this it's own function early on for easier access to list of comps
+    comp_set_dict = [row for row in element_pmf_results if row['MOF'] == mof_list[0]]
+    # comps_array = []
+    # for row in comp_set_dict:
+    #     comps_array.append([float(row[gas]) for gas in gases])
+    comps_array = np.array([[float(row[gas]) for gas in gases] for row in comp_set_dict])
+    # Figure out what is different between commented approach and current one!
+
+    # Determine the set of points used to create bins
+    bin_points = []
+    for i in range(len(gases)):
+        lower_comp = min(comps_array[:,i])
+        upper_comp = max(comps_array[:,i])
+        # Brian Bins
+        lower_lim = lower_comp - 0.5*(upper_comp-lower_comp)/(num_bins-1)
+        upper_lim = upper_comp + 0.5*(upper_comp-lower_comp)/(num_bins-1)
+        # Jenna Bins
+        # lower_lim = lower_comp
+        # upper_lim = upper_comp + (upper_comp-lower_comp)/(num_bins)
+        bin_points_temp = np.linspace(lower_lim, upper_lim, num=num_bins+1, endpoint=True)
+        # Extra steps to deal with impercision of values generated by np.linspace
```

Comment:
I had noticed that about 9-12 digits out I would see non-zero digits (i.e. 0.3000000000007 rather than 0.30). Although this should be fine when my bin spacing is halfway between the relevant compositions, which they currently are, when a bin and a bin point fall on the exact same value, which they might if I reintroduce interpolation or have some reason to use larger or smaller bins, the mix between bin values like 0.300000007 and 0.30 would make using if x >= y or (<= or ==) shift certain points into bins where I did not want them. I can see if there is a way to specify the precision of np.linspace, or just adjust the precision of an array, but this was just the fix I came up with first.

Reply:
Are you familiar with how floating point numbers are stored? You shouldn't expect to have even boundaries; they will always be a very, very tiny fraction off from the "correct" number. It shouldn't rarely, rarely, rarely (if ever) matter though. If you adjust the precision, you will just have the errors be more digits out, since floating point numbers are not exact representations. But it really shouldn't matter though. Why does it matter if a 0.30...0x value goes into the <0.3 bin or the >= 0.3 bin? Isn't it all the same?